Add layout analysis and distribution pass

harsh-nod opened the pull request:
This patch adds a pass to convert
transfer_read + contract + transfer_write to
loads + mma.sync + store. It does so by
introducing an implicit high-dimensional layout (8D) that consists of batch, laneId and vecId dimensions.

harsh-nod committed:
Add layout analysis and distribution pass

This patch adds a pass to convert
transfer_read + contract + transfer_write to
loads + mma.sync + store. It does so by
introducing an implicit high-dimensional layout (8D)
that consists of batch, laneId and vecId dimensions.

nicolasvasilache reviewed (commented):
It's great to see the higher-D vector idea being prototype, we had been discussing this for at least 2 years :)

Thanks a lot Harsh for pushing on this, can you please send this upstream ?
This is not at all IREE specific.

Thanks much!

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Passes.td:
```diff
@@ -474,6 +474,12 @@ def ConvertToNVVM :
   let constructor = "mlir::iree_compiler::createConvertToNVVMPass()";
 }
 
+def LLVMGPULayoutAnalysisAndDistribution :
```
Can we make it a transform dialect op instead of a pass? That might be easier to push some early execution test to validate that things are correct.
This would work the same as a pass as it would run on the whole function.

harsh-nod replied:
Sounds good. Will change.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,517 @@
+  // Contains the order of layout dims for each of the tensor dims
+  OrderArray order;
+  // Shape of the tensor when used in mma
+  std::array<int, maxTensorDims> canonicalShape;
```
is it meant to be maxVectorDim? There should be no more tensor when doing this conversion?

harsh-nod replied:
So for the mma.m16n8k16 case, the A-matrix has a canonical shape 16x16, B has 16x8 and C has 16x8. So that's what I was referring to here as maxTensorDims (=2). But maybe this is confusing, should I rename maxTensorDims to maxVectorDims?

harsh-nod committed:
Convert pass to transform dialect op

Update test to include schedule as well.
Also removed include file that was causing compile error.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,499 @@
+                          ArrayRef<int64_t> cShape) {
+  if ((aShape[0] % 16 == 0) && (aShape[1] % 16 == 0) && (cShape[0] % 16 == 0) &&
+      (cShape[1] % 8 == 0)) {
+    // B-matrix could be transposed
```
I don't think you want to handle the transposition at this stage. The contract can be transformed to the canonical form before doing the SIMD to SIMT, this is probably a better IR to start with, then the transposition will happen in the transfer read in this case.
I would remove it and start from an IR where the contract is already doing `a*t(b)`

harsh-nod replied:
That makes a lot of sense. Thanks will remove that condition from the if statement.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,499 @@
+    if (auto readOp = matrix.getDefiningOp<vector::TransferReadOp>()) {
+      Value src = readOp.getSource();
+      layoutMap.try_emplace(src, layout);
+    }
```
I don't understand that part, the source of transfer_read is a memref, it shouldn't have a layout. Layout should only be attached to vector values.
You shouldn't' need any propagation in this case?

harsh-nod replied:
Good point! Yes, the memref should not have a layout. I will remove this.

ThomasRaoux reviewed (commented):
Looks good, couple more comments/questions

harsh-nod committed:
Remove handling of mma with B transpose and do not propagate
to transfer reads

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,485 @@
+}
+
+std::string Layout::str() const {
+  std::stringstream ss;
```
can you convert that to LLVM style of printing out debug info.
Example:
https://github.com/openxla/iree/blob/624db6dc2be50b64c519353bdbc8c5dc8dce3280/compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp#L37

harsh-nod replied:
sure will change.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,485 @@
+  rank = dimOrder.size();
+}
+
+void Layout::reduceDim(int i) {
```
this seems unused.

harsh-nod replied:
yes its for the multi-dim reductions but will remove for now.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp:
```diff
@@ -0,0 +1,485 @@
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
```
please add IREE copyright header

harsh-nod replied:
will do.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/layout_analysis_and_distribution.mlir:
```diff
@@ -0,0 +1,120 @@
+    memref.assume_alignment %2, 64 : memref<16x8xf16>
+    %3 = vector.transfer_read %0[%c0, %c0], %cst_0 {in_bounds = [true, true]} : memref<16x16xf16>, vector<16x16xf16>
+    %4 = vector.transfer_read %1[%c0, %c0], %cst_0 {in_bounds = [true, true]} : memref<16x8xf16>, vector<16x8xf16>
+    %5 = vector.contract {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %3, %4, %cst : vector<16x16xf16>, vector<16x8xf16> into vector<16x8xf16>
```
I would expect the contract to have the maps:
`(d0, d1, d2) -> (d0, d2), (d0, d1, d2) -> (d1, d2), affine_map<(d0, d1, d2) -> (d0, d1)`
Otherwise this won't map to a mma.sync.

harsh-nod replied:
I could be understanding this wrong but based on https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-instructions-mma,  it says ```Perform a MxNxK matrix multiply and accumulate operation, D = A*B+C, where the A matrix is MxK, the B matrix is KxN, and the C and D matrices are MxN.``` Doesn't that mean that the B matrix is KxN and so should be ```(d0, d1, d2) -> (d2, d1)``` or am I missing something?

ThomasRaoux replied:
This part is a bit confusing, it does say B is `KxN` but I believe it uses the `row.col` layout meaning B is in col major layout. From what I have seen mma.sync really represents a matmul with transpose b so it is likely going to be much simpler to represent it that way. I think your solution works as well as it represent the layout transposed and basically the work of transposing is happening when loading from memref.
We can figure that out later. If we can have an execution test based on this in one of the next step it could be interesting to play around with this.

harsh-nod replied:
Ah that makes sense. I missed the part about A being row major and B being column major. I will add an end to end test in a subsequent patch.

ThomasRaoux reviewed (commented):
Few more comments

harsh-nod committed:
Address Thomas' comments

- Change debug to LLVM style
- Add IREE copyright header
- Remove reduceDim function

harsh-nod committed:
Add e2e test using layout pass

ThomasRaoux commented on tests/transform_dialect/cuda/mma_using_layout_analysis.mlir:
```diff
@@ -0,0 +1,20 @@
+// RUN: iree-compile %s --iree-hal-target-backends=cuda \
+// RUN:     --iree-codegen-llvmgpu-enable-transform-dialect-jit=false \
+// RUN:     --iree-codegen-llvmgpu-use-transform-dialect=%p/reduction_v2_codegen_spec.mlir | \
+// RUN: iree-run-module --function=matmul --device=cuda --input="16x16xf16=1" --input="16x8xf16=2" |\
```
Should we have more complex inputs to make sure we are reading the right elements? We won't be able to tell if we are loading things right with this case?

harsh-nod replied:
Sure I can increase the tensor sizes.

ThomasRaoux replied:
I meant keep the size but have non uniform input?

harsh-nod replied:
Oh I see, yes sure. I can add non-uniform input.

harsh-nod replied:
So what's the best way to add non-uniform input? I think numpy files would be ideal here but not sure how to integrate them with the BUILD script. Or do you have any other ideas?

harsh-nod replied:
nvm I think I can just paste them in

ThomasRaoux replied:
here is an example:
https://github.com/openxla/iree/blob/817c771c783669d7a5cfaa2b1199f3ed58b1c305/tests/e2e/regression/transpose.mlir#L18

harsh-nod replied:
Thanks! I have updated the patch with a non uniform test. Please merge when you can as I don't have merge privileges.

harsh-nod committed:
Increase the tensor sizes in e2e test

harsh-nod committed:
Revert back to original sizes, still uniform input,
but should pass GPU tests

harsh-nod committed:
Add non-uniform test

Also fix bug so that test passes.
